src/org/pentaho/reporting/platform/plugin/cache/FileSystemCacheBackend.java

```diff
@@ -51,50 +51,82 @@ public void setCachePath( final String cachePath ) {
 
   @Override
   public boolean write( final List<String> key, final Serializable value ) {
-    final File file = new File( cachePath + StringUtils.join( cleanKey( key ), File.separator ) );
-
-    final ObjectOutputStream oos;
-    final FileOutputStream fout;
+    final String filePath = cachePath + StringUtils.join( cleanKey( key ), File.separator );
+    final File file = new File( filePath );
+    lock( filePath );
     try {
+      //create file structure
       file.getParentFile().mkdirs();
       if ( !file.exists() ) {
         file.createNewFile();
       }
-
-      fout = new FileOutputStream( file );
-      oos = new ObjectOutputStream( fout );
-      oos.writeObject( value );
-      oos.close();
-      fout.close();
+      //closable resources
+      try (
+        final FileOutputStream fout = new FileOutputStream( file );
+        final ObjectOutputStream oos = new ObjectOutputStream( fout ) ) {
+        oos.writeObject( value );
+      }
     } catch ( final IOException e ) {
       logger.error( "Can't write cache: ", e );
       return false;
+    } finally {
+      unlock( filePath );
     }
-
     return true;
+
+  }
+
+  private void lock( final String filePath ) {
+    final File lock = new File( filePath + ".lock" );
+    while ( lock.exists() ) {
```

Comment:
File locks are held on behalf of the entire Java virtual machine. They are not suitable for controlling access to a file by multiple threads within the same virtual machine.

https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileLock.html

@tmorgner